feat: Provide New Relic scaler

marcelobartsch opened the pull request:
Provide New Relic scaler.

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)* https://github.com/kedacore/keda-docs/pull/591
- [x] Changelog has been updated

Relates to https://github.com/kedacore/keda/issues/2290

This is a 'redo' of PR https://github.com/kedacore/keda/pull/2286 as I had an issue in my REPO and had to start over, but I manage to rescue the files.

JorTurFer commented:
hi @marcelobartsch 
What do we need to set up in order to run e2e tests? I mean, AFAIK New Relic has a free tier that should be enough for our use case. Am I right?

marcelobartsch commented:
> hi @marcelobartsch What do we need to set up in order to run e2e tests? I mean, AFAIK New Relic has a free tier that should be enough for our use case. Am I right?

Yes, in fact all my tests have been done using a free account tier, so I don't mix my test with our company data, so it is perfectly good!

marcelobartsch commented:
> LGTM in general, but there is a pending change to make e2e test possible. Please, update the (GitHub) workloads to pass the new environment variables to the actions where tests are executed. There are 3 places to update:
> 
>     * https://github.com/kedacore/keda/blob/main/.github/workflows/main-build.yml#L64-L82
> 
>     * https://github.com/kedacore/keda/blob/main/.github/workflows/nightly-e2e.yml#L18-L36
> 
>     * https://github.com/kedacore/keda/blob/main/.github/workflows/pr-e2e.yml#L62-L84
> 
> 
> You should add these lines:
> 
> ```yaml
>     NEWRELIC_API_KEY: ${{ secrets.NEWRELIC_API_KEY}}
>     NEWRELIC_ACCOUNT_ID: ${{ secrets.NEWRELIC_ACCOUNT_ID}}
> ```
> 
> I have just added them in the repository. After your change, I will trigger e2e test to check it :)
> 
> Thanks for this contribution! ❤️

For the E2E you will also need the NEWRELIC_LICENSE environment variable, the scaler itself don't need it, but the nri-bundle which install the components needed in the kubernetes cluster use it.

NEWRELIC_API_KEY always start with NRAK
NEWRELIC_LICENSE always ENDs with NRAL

This is a common issue , at least for us , we always mix up them when we started years ago :)

I added those lines to the files you mentions

```yaml
NEWRELIC_ACCOUNT_ID: ${{ secrets.NEWRELIC_ACCOUNT_ID}}
NEWRELIC_API_KEY: ${{ secrets.NEWRELIC_API_KEY}}
NEWRELIC_LICENSE: ${{ secrets.NEWRELIC_LICENSE}}
```

NEWRELIC_LICENSE is new so probably you will need to add it to secrets.

JorTurFer commented:
> > LGTM in general, but there is a pending change to make e2e test possible. Please, update the (GitHub) workloads to pass the new environment variables to the actions where tests are executed. There are 3 places to update:
> > ```
> > * https://github.com/kedacore/keda/blob/main/.github/workflows/main-build.yml#L64-L82
> > 
> > * https://github.com/kedacore/keda/blob/main/.github/workflows/nightly-e2e.yml#L18-L36
> > 
> > * https://github.com/kedacore/keda/blob/main/.github/workflows/pr-e2e.yml#L62-L84
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You should add these lines:
> > ```yaml
> >     NEWRELIC_API_KEY: ${{ secrets.NEWRELIC_API_KEY}}
> >     NEWRELIC_ACCOUNT_ID: ${{ secrets.NEWRELIC_ACCOUNT_ID}}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have just added them in the repository. After your change, I will trigger e2e test to check it :)
> > Thanks for this contribution! ❤️
> 
> For the E2E you will also need the NEWRELIC_LICENSE environment variable, the scaler itself don't need it, but the nri-bundle which install the components needed in the kubernetes cluster use it.
> 
> NEWRELIC_API_KEY always start with NRAK NEWRELIC_LICENSE always ENDs with NRAL
> 
> This is a common issue , at least for us , we always mix up them when we started years ago :)
> 
> I added those lines to the files you mentions
> 
> ```yaml
> NEWRELIC_ACCOUNT_ID: ${{ secrets.NEWRELIC_ACCOUNT_ID}}
> NEWRELIC_API_KEY: ${{ secrets.NEWRELIC_API_KEY}}
> NEWRELIC_LICENSE: ${{ secrets.NEWRELIC_LICENSE}}
> ```
> 
> NEWRELIC_LICENSE is new so probably you will need to add it to secrets.

I will add it right now, thanks :)

JorTurFer commented:
The terms are a bit tricky, but I have updated both following your instructions :) 
Thanks

JorTurFer commented:
/run-e2e new-relic.test*

JorTurFer commented:
/run-e2e new-relic.test*

zroubalik commented:
/run-e2e new-relic.test*

marcelobartsch commented:
Woking on fix the E2E test, for some reason it works on local (which I know is no excuse :D )

marcelobartsch commented:
@JorTurFer any chance the e2e test had limited from where it can download the helm charts? I got this error

https://github.com/kedacore/keda/runs/4448161285?check_suite_focus=true#step:8:96

JorTurFer commented:
> @JorTurFer any chance the e2e test had limited from where it can download the helm charts? I got this error
> 
> ```
> Error: failed to download "newrelic/nri-bundle"
> ```
> 
> [kedacore/keda/runs/4448161285?check_suite_focus=true#step:8:96](https://github.com/kedacore/keda/runs/4448161285?check_suite_focus=true#step:8:96)

I think that it's not. I will try your PR in my local

JorTurFer commented on tests/scalers/new-relic.test.ts:
```diff
@@ -0,0 +1,257 @@
+  if (!newRelicAccountId) {
+    t.fail('NEWRELIC_ACCOUNT_ID environment variable is required for newrelic tests tests')
+  }
+  sh.exec(`kubectl create namespace ${newRelicNamespace}`)
+  sh.exec(`helm repo add ${newRelicRepoName} ${newRelicRepoUrl}`)
+
+  let helmInstallStatus = sh.exec(`helm upgrade --install --set global.cluster=${kuberneteClusterName} --set prometheus.enabled=true --set ksm.enabled=true --set global.lowDataMode=true --set global.licenseKey=${newRelicLicenseKey} --timeout 600s --set logging.enabled=false --set ksm.enabled=true --set logging.enabled=true --namespace ${newRelicNamespace} nri-keda ${newRelicHelmPackageName}`).cod ...
```
Probably helm repo update should be here xD

marcelobartsch replied:
🤦🏼 I think you are totally right xD 
Added to the test file.

JorTurFer commented:
/run-e2e new-relic.test*

JorTurFer commented on tests/scalers/new-relic.test.ts:
```diff
@@ -0,0 +1,257 @@
+const newRelicNamespace = 'new-relic'
+const newRelicRepoUrl = 'https://helm-charts.newrelic.com'
+const newRelicRepoName = 'new-relic'
+const newRelicHelmPackageName = 'newrelic/nri-bundle'
```
```suggestion
const newRelicHelmPackageName = 'nri-bundle'
```

marcelobartsch replied:
done!

JorTurFer commented on tests/scalers/new-relic.test.ts:
```diff
@@ -0,0 +1,257 @@
+  sh.exec(`kubectl create namespace ${newRelicNamespace}`)
+  sh.exec(`helm repo add ${newRelicRepoName} ${newRelicRepoUrl}`)
+  sh.exec(`helm repo update`)
+  let helmInstallStatus = sh.exec(`helm upgrade --install --set global.cluster=${kuberneteClusterName} --set prometheus.enabled=true --set ksm.enabled=true --set global.lowDataMode=true --set global.licenseKey=${newRelicLicenseKey} --timeout 600s --set logging.enabled=false --set ksm.enabled=true --set logging.enabled=true --namespace ${newRelicNamespace} nri-keda ${newRelicHelmPackageName}`).cod ...
```
```suggestion
  let helmInstallStatus = sh.exec(`helm upgrade --install --set global.cluster=${kuberneteClusterName} --set prometheus.enabled=true --set ksm.enabled=true --set global.lowDataMode=true --set global.licenseKey=${newRelicLicenseKey} --timeout 600s --set logging.enabled=false --set ksm.enabled=true --set logging.enabled=true --namespace ${newRelicNamespace} nri-keda ${newRelicRepoName}/${newRelicHelmPackageName}`).code
```

marcelobartsch replied:
Fixed! thanks I was not seeing that error and was a 'dummy' error :D

JorTurFer reviewed (changes requested):
There is a typo in `newRelicHelmPackageName`. Look the suggestion to see where is and how it could be fixed

JorTurFer commented on tests/scalers/new-relic.test.ts:
```diff
@@ -0,0 +1,257 @@
+  - type: new-relic
+    metadata:
+      account: '{{NEWRELIC_ACCOUNT_ID}}'
+      threshold: '100'
```
```suggestion
      region: 'EU'
      threshold: '100'
```

marcelobartsch replied:
Fixed, I added as a variable, so local test can be run on either US or EU region, but by default will use EU region

JorTurFer reviewed (changes requested):
Please, update the e2e test to use EU region because our new relic account is in EU. Thanks :)

marcelobartsch commented:
/run-e2e new-relic.test*

JorTurFer commented:
/run-e2e new-relic.test*

JorTurFer reviewed (approved):
LGTM!
Thanks fot your contribution :heart:

JorTurFer reviewed (changes requested):
Please, update CHANGELOG.md

JorTurFer reviewed (approved):
Now yes, LGTM :)

zroubalik commented on pkg/scalers/newrelic_scaler.go:
```diff
@@ -0,0 +1,205 @@
+	"log"
+	"strconv"
+
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	v2beta2 "k8s.io/api/autoscaling/v2beta2"
+	"k8s.io/apimachinery/pkg/labels"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/newrelic/newrelic-client-go/newrelic"
+	"github.com/newrelic/newrelic-client-go/pkg/nrdb"
```
```suggestion
	"github.com/newrelic/newrelic-client-go/newrelic"
	"github.com/newrelic/newrelic-client-go/pkg/nrdb"
	v2beta2 "k8s.io/api/autoscaling/v2beta2"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/metrics/pkg/apis/external_metrics"
	logf "sigs.k8s.io/controller-runtime/pkg/log"

	kedautil "github.com/kedacore/keda/v2/pkg/util"
```

Just a nit, we tend to group all general imports and keep KEDA related in a separate block

zroubalik commented on pkg/scalers/newrelic_scaler.go:
```diff
@@ -0,0 +1,205 @@
+		return nil, fmt.Errorf("no %s given", queryKey)
+	}
+
+	if val, ok := config.TriggerMetadata[metricName]; ok && val != "" {
```
I think we can drop this one, see following discussion:
https://github.com/kedacore/keda/pull/2354#discussion_r758090060

marcelobartsch replied:
@zroubalik I don't understand what you want to be dropped from here.

zroubalik commented on pkg/scalers/newrelic_scaler.go:
```diff
@@ -0,0 +1,205 @@
+
+func (s *newrelicScaler) GetMetricSpecForScaling(context.Context) []v2beta2.MetricSpec {
+	targetMetricValue := resource.NewQuantity(int64(s.metadata.threshold), resource.DecimalSI)
+	var metricName string
+	if s.metadata.metricName == "" {
+		// TODO: Find a better way to generate a small string based on the nrql string, or maybe just use UUID?
+		hash := md5.Sum([]byte(s.metadata.nrql))
+		metricName = kedautil.NormalizeString(fmt.Sprintf("new-relic-%s", hex.EncodeToString(hash[:])))
+	} else {
+		metricName = kedautil.NormalizeString(fmt.Sprintf("new-relic-%s", s.metadata.metricName))
+	}
```
You can drop this whole section and just use `GenerateMetricNameWithIndex(s.metadata.scalerIndex, metricName)`, it ensures that the metric name is unique.

marcelobartsch replied:
Ok, so we make metricName a requiered field on the metadata? is that right?

zroubalik replied:
No, it's the other way round :)

We shoulnd't ask for metricName, generating one internally is enough.

zroubalik commented on pkg/scalers/newrelic_scaler.go:
```diff
@@ -0,0 +1,205 @@
+		return nil, fmt.Errorf("missing %s value", threshold)
+	}
+
+	// If Query Return an Empty Data , shall we treat it as an error or not
+	// default is NO error is returned when query result is empty/no data
+	if val, ok := config.TriggerMetadata[noDataErr]; ok {
+		noDataErr, err := strconv.ParseBool(val)
+		if err != nil {
+			return nil, fmt.Errorf("noDataErr has invalid value")
+		}
+		meta.noDataErr = noDataErr
+	} else {
+		meta.noDataErr = false
+	}
```
I don't see this in documentation

zroubalik reviewed (commented):
Generally looking good, I have some minor comments.

marcelobartsch commented:
/run-e2e new-relic.test*

JorTurFer commented:
/run-e2e new-relic.test*

zroubalik reviewed (approved):
LGTM, thanks!

zroubalik commented:
/run-e2e new-relic.test*

zroubalik commented on pkg/scalers/newrelic_scaler_test.go:
```diff
@@ -0,0 +1,82 @@
+	// all properly formed
+	{map[string]string{"account": "0", "region": "EU", "threshold": "100", "queryKey": "somekey", "nrql": "SELECT average(cpuUsedCores) as result FROM K8sContainerSample WHERE containerName='coredns'"}, false},
+	// account as String
+	{map[string]string{"account": "ABC", "metricName": "results", "threshold": "100", "queryKey": "somekey", "nrql": "SELECT average(cpuUsedCores) as result FROM K8sContainerSample WHERE containerName='coredns'"}, true},
```
metric name is still in the test

zroubalik reviewed (commented):
Please remove `metricName` from the test and fix the conflict in the workflows & Changelog and we can merge this!

Thanks
